src/wrapper.js

```diff
@@ -0,0 +1,162 @@
+  NextStepShown: 'NextStepShown',
+  PlayerInitComplete: 'PlayerInitComplete',
+  SurveyPlayed: 'SurveyPlayed',
+  SurveyCompleted: 'SurveyCompleted',
+  TaskCompleted: 'TaskCompleted',
+  TasksLoaded: 'TasksLoaded',
+  UserStoppedWalkthru: 'UserStoppedWalkthru',
+  UserStoppedWalkthruAfterStop: 'UserStoppedWalkthruAfterStop',
+  WalkthruAborted: 'WalkthruAborted',
+  WalkthruFailedToStart: 'WalkthruFailedToStart',
+  WalkthruSelected: 'WalkthruSelected'
+};
+
+let afterApiLoad = function afterApiLoad (fn) {
+  if (!isLoadedPromise) {
+    isLoadedPromise = new Promise(function isLoaded (resolve, reject) {
+      let count = 1;
+      let interval = setInterval(() => {
+        if (typeof WalkMeAPI !== 'undefined') {
+          resolve(true);
+          clearInterval(interval);
+        } else if (count === API_LOAD_ATTEMPTS) {
+          reject(new Error(false));
+          clearInterval(interval);
+        }
+        count++;
+      }, API_LOAD_INTERVAL);
+    });
+  }
+
+  return function isLoaded () {
+    let args = arguments;
+    return isLoadedPromise.then(() => fn.apply(null, args)).catch(() => false);
+  };
+};
+
+wrapper.runSnippet = function runSnippet (walkmeSnippetUrl) {
+  let walkme = document.createElement('script');
+  let script;
+  walkme.type = 'text/javascript';
+  walkme.async = true;
+  walkme.src = walkmeSnippetUrl;
+  script = document.getElementsByTagName('script')[0];
+  script.parentNode.insertBefore(walkme, script);
+  window._walkmeConfig = { smartLoad: true };
+};
+
+wrapper.startWalkthrough = afterApiLoad(function (id) {
+  WalkMeAPI.startWalkthruById(id);
+});
+
+wrapper.isLoaded = afterApiLoad(() => true);
+
+let eventSubscribers = Object.keys(wrapper.Event).reduce((subscribers, event) => {
+  subscribers[event] = [];
+  return subscribers;
+}, {});
+
+/**
+ * Listen to a WalkMe event
+ *
+ * @param  {String} eventName - the WalkMe event
+ * @param  {Function} handler - the method to execute when this event is received
+ */
+wrapper.on = afterApiLoad(function (eventName, handler) {
+  window.menu_event = window.menu_event || function onMenuEvent (eventData) {
+    eventSubscribers[eventData.Type].forEach(subscriber => {
+      subscriber(eventData);
+    });
+  };
+
+  window.walkme_event = window.walkme_event || function onWalkMeEvent (eventData) {
+    eventSubscribers[eventData.Type].forEach(subscriber => {
+      subscriber(eventData);
+    });
+  };
+
+  eventSubscribers[eventName].push(handler);
+});
+
+const lastStepMemo = {};
+function isLastStep (walkthruId, stepIndex) {
+  if (walkthruId in lastStepMemo) {
+    return lastStepMemo[walkthruId] === stepIndex;
+  }
+
+  let isUndocumentedApiAvailable = _walkMe && typeof _walkMe.getTutorials === 'function';
+
+  if (isUndocumentedApiAvailable) {
+    let tutorial = find(_walkMe.getTutorials(), {Id: walkthruId});
+    let sortedSteps = sortBy(tutorial.Steps, 'StepIndex');
+    let lastStep = tutorial && findLast(sortedSteps, step => {
```

Comment:
Does tutorial/sortedSteps/lastStep change depending on where the user is? Is there a way get lastStep.StepIndex earlier on instead of every time in the isLastStep function?

Reply:
there isn't an earlier place to get it because we don't know that we care until they start completing a walkthrough

but we are caching the result in `lastStepMemo` so we're not executing this every time necessarily